### PR TITLE
Operator: public authorized senders

### PR DESCRIPTION
**Pull request by alexroan**

### CODE REVIEW

File: evm-contracts/src/v0.7/dev/Operator.sol [hunk 1]
```diff
@@ -27,6 +27,11 @@ contract Operator is
     uint8 dataVersion;
   }
 
+  struct Authorization {
+    bool authorized;
+    uint256 index;
```
Comment (se3000) [inline evm-contracts/src/v0.7/dev/Operator.sol:32]: You can save a word of storage and a bunch of gas by making this a `uint255`, but realistically it probably doesn't need to be any bigger than a `uint8`.

**Review by se3000 — CHANGES_REQUESTED** [review]
Overall looks great, couple small things, but this delivers on the story criteria. One thing that is outside the scope of the story, but we should probably do is move this to be `setAuthorizedSenders` instead of adding each one individually. This allows us to declaratively set permissions, which is easier to reason about and make quick changes than diffs for each addition and removal of a sender.

Whether you want to do that in this PR or another is your call.

**Comment by alexroan** [comment]
@se3000 Happy to implement `setAuthorizedSenders(address[] calldata nodes, bool[] calldata permissions)` in this PR. Would the purpose of that function be to:

1. Overwrite existing authorized senders? (in which case, we won't need the `permissions` parameter and assume that every address in the parameter should have permissions.)
2. Act on existing configuration? (More complicated, as it would need to loop through each, check against existing config, and alter accordingly)

1 for simplicity, 2 for adaptability. What do you think?

**Review by se3000 — CHANGES_REQUESTED** [review]
We should also emit a log announcing the new authorized sender list.

File: evm-contracts/src/v0.7/dev/Operator.sol [hunk 2]
```diff
@@ -213,29 +218,48 @@ contract Operator is
...
-  function setAuthorizedSender(address node, bool allowed)
+  function setAuthorizedSenders(address[] calldata senders)
     external
     override
     onlyOwner()
   {
-    s_authorizedSenders[node] = allowed;
+    require(senders.length > 0, "Must have at least 1 authorized sender");
+    // Set previous authorized senders to false
+    for (uint256 i = 0; i < s_authorizedSenderList.length; i++) {
```
Comment (se3000) [inline evm-contracts/src/v0.7/dev/Operator.sol:244]: I am not sure if `s_authorizedSenderList.length` reads from storage every time. Might be cheaper to cache the value in a local variable. Can you write a test, possibly throwaway to check that? Can throwaway the test, just want to check.

Comment (alexroan) [reply evm-contracts/src/v0.7/dev/Operator.sol:244]: Great spot! 

On setting 3 authorized senders, caching the length saves 795 gas!

File: evm-contracts/src/v0.7/dev/Operator.sol [hunk 3]
```diff
@@ -213,29 +218,49 @@ contract Operator is
...
+    for (uint256 i = 0; i < authorizedSendersLength; i++) {
+      s_authorizedSenders[s_authorizedSenderList[i]] = false;
+    }
+    // Set new to true
+    for (uint256 i = 0; i < senders.length; i++) {
+      s_authorizedSenders[senders[i]] = true;
+    }
+    // Replace list
+    s_authorizedSenderList = senders;
+    emit AuthorizedSendersChanged(s_authorizedSenderList);
```
Comment (se3000) [inline evm-contracts/src/v0.7/dev/Operator.sol:254]:
```suggestion
    emit AuthorizedSendersChanged(senders);
```

File: evm-contracts/src/v0.7/dev/Operator.sol [hunk 4]
```diff
@@ -213,29 +218,49 @@ contract Operator is
...
+  }
+
+  /**
+   * @notice Retrieve a list of authorized senders
+   * @return array of addresses
+   */
+  function getAuthorizedSenders() external view override returns (address[] memory) {
+    address[] memory authorizedSendersList = s_authorizedSenderList;
+    return authorizedSendersList;
```
Comment (se3000) [inline evm-contracts/src/v0.7/dev/Operator.sol:263]:
```suggestion
    return s_authorizedSenderList;
```

**Review by se3000 — CHANGES_REQUESTED** [review]

File: evm-contracts/src/v0.7/dev/Operator.sol (hunk 4 above)
Comment (alexroan) [reply evm-contracts/src/v0.7/dev/Operator.sol:263]: 🤔 wtf was I doing here?

**Comment by alexroan** [comment]
Squashed history a bit since it was getting long.

**Review by se3000 — APPROVED** [review]

**Pull request merged**